src/transformers/models/ict/modeling_ict.py

```diff
@@ -0,0 +1,854 @@
+        )
+        style_loss += self.criterion(
+            self.compute_gram_matrix(x_vgg["relu4_4"]), self.compute_gram_matrix(y_vgg["relu4_4"])
+        )
+        style_loss += self.criterion(
+            self.compute_gram_matrix(x_vgg["relu5_2"]), self.compute_gram_matrix(y_vgg["relu5_2"])
+        )
+
+        return style_loss
+
+
+class IctPerceptualLoss(nn.Module):
+    r"""
+    Perceptual loss, VGG-based https://arxiv.org/abs/1603.08155
+    https://github.com/dxyang/StyleTransfer/blob/master/utils.py
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.vgg = VGG19()
+        self.criterion = torch.nn.L1Loss()
+        self.weights = [1.0, 1.0, 1.0, 1.0, 1.0]
+
+    def forward(self, x, y):
+        # Compute features
+        x_vgg, y_vgg = self.vgg(x), self.vgg(y)
+
+        content_loss = 0.0
+        content_loss += self.weights[0] * self.criterion(x_vgg["relu1_1"], y_vgg["relu1_1"])
+        content_loss += self.weights[1] * self.criterion(x_vgg["relu2_1"], y_vgg["relu2_1"])
+        content_loss += self.weights[2] * self.criterion(x_vgg["relu3_1"], y_vgg["relu3_1"])
+        content_loss += self.weights[3] * self.criterion(x_vgg["relu4_1"], y_vgg["relu4_1"])
+        content_loss += self.weights[4] * self.criterion(x_vgg["relu5_1"], y_vgg["relu5_1"])
+
+        return content_loss
+
+
+class IctGuidedUpsampler(IctPretrainedModel):
+    def __init__(self, config: IctConfig):
+        super().__init__(config)
+
+        self.generator = IctInpaintGenerator(config)
+        self.adversarial_loss = IctAdversarialLoss(config)
+        self.l1_loss = nn.L1Loss()
+        self.style_loss = IctStyleLoss()
+        self.perceptual_loss = IctPerceptualLoss()
+        self.output_image_size = config.output_image_size
+
+        self.post_init()
+
+    # modified from https://github.com/raywzy/ICT/blob/59dd12d374d47cdf0dce90923017ca3657e6aa0b/Guided_Upsample/src/dataset_my.py#L203-L209
+    # and https://github.com/raywzy/ICT/blob/59dd12d374d47cdf0dce90923017ca3657e6aa0b/Guided_Upsample/src/dataset_my.py#L183-L186
+    def resize(self, img: torch.Tensor, target_height: int, target_width: int):
+        img = img.to(self.device)
+        # If the image tensor is in the format (N, H, W, C), change it to (N, C, H, W)
+        if img.dim() == 4 and img.shape[1] > img.shape[3]:
+            img = img.permute(0, 3, 1, 2)
+
+        # Handle boolean tensors
+        if img.dim() == 3:
+            img = img.unsqueeze(1)
+
+        # Center crop for non-square images
+        _, _, height, width = img.shape
+        if height != width:
+            side_length = min(height, width)
+            height_offset = (height - side_length) // 2
+            width_offset = (width - side_length) // 2
+            img = img[:, :, height_offset : height_offset + side_length, width_offset : width_offset + side_length]
+
+        img = img.float()
+        img = F.interpolate(img, size=(target_height, target_width), mode="bicubic")
+
```

Comment:
@amyeroberts  This code used to convert from torch -> numpy -> PIL but it was refactored to just do the resizing with torch. But it seems that it's failing tests that test for determinism. My guess is that it's something to do with `interpolate` but couldn't figure it out.

Reply:
Could you specify which tests are failing? 

Doing torch -> numpy -> PIL is quite funny within the model, especially when torch already can upsample and doing so will remove gradient information. Is that pattern from the original codebase? 

Why do you suspect that it's coming from `interpolate`?

Reply:
Yup, that was the pattern from the original codebase (from [here](https://github.com/raywzy/ICT/blob/59dd12d374d47cdf0dce90923017ca3657e6aa0b/Guided_Upsample/src/dataset_my.py#L203-L209) and [here](https://github.com/raywzy/ICT/blob/59dd12d374d47cdf0dce90923017ca3657e6aa0b/Guided_Upsample/src/dataset_my.py#L183-L186)). 

Here are the five failings tests separated by causes.

1. `IctModelTest.test_determinism`
2. `IctModelTest.test_model_outputs_equivalence`
3. `IctModelTest.test_save_load`
Never mind about interpolate. I was debugging it and saw that appearance_prior was changing and realized that it was coming from `pred = torch.multinomial(probs, num_samples=1)`

But I set `torch.manual_seed(3)` in `test_modeling_ict.py` and it seems to be not working...?

4. `IctModelTest.test_retain_grad_hidden_states_attentions`
I'm getting the error below but I still haven't been able to figure out
```
self.assertIsNotNone(hidden_states.grad) 
AssertionError: unexpectedly None
```


5. `IctModelIntegrationTest.test_inference_masked_image_modeling`
My guess is that I need to run the original model with the same image and see if the tensors match? If the original model also uses randomization within the code (not just the weights), how should I go about comparing these?

Reply:
Another test that failing is `build / build_pr_documentation` which is saying:
```
The docstring of IctModel.forward comports the following issue(s) and needs 
fixing:
- The return block is empty.
```

which seems to be pointing at this part:
```python
def forward(
        self,
        pixel_values: Optional[torch.Tensor],
        bool_masked_pos: Optional[torch.BoolTensor] = None,
        clusters: Optional[torch.Tensor] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = None,
        return_dict: Optional[bool] = None,
    ) -> Union[Tuple, MaskedImageModelingOutput]:
        r"""
        Returns:

        Example:
        ```python
        >>> import torch
        >>> import numpy as np
        >>> from PIL import Image
        >>> import requests
```

But I was looking at other `forward` methods of other models and it seems that `Returns:` are empty, so I left it as it is. But maybe I'm missing something.

Reply:
Also, [this tells me that there's a style error](https://app.circleci.com/pipelines/github/huggingface/transformers/67643/workflows/3f1cd6cc-d0d0-4341-98c3-61ad79716542/jobs/845594/parallel-runs/0/steps/0-114) somewhere but not telling me where exactly 😅